math/src/field/smallprimefield/mod.rs

```diff
@@ -0,0 +1,493 @@
+            domain_size.is_power_of_two(),
+            "domain size must be a power of 2"
+        );
+        let root = Self::get_root_of_unity(domain_size.try_into().unwrap());
+        let mut twiddles = Self::get_power_series(root, domain_size / 2);
+        fft::permute(&mut twiddles);
+        twiddles
+    }
+
+    pub unsafe fn get_inv_twiddles(domain_size: usize) -> Vec<Self> {
+        debug_assert!(
+            domain_size.is_power_of_two(),
+            "domain size must be a power of 2"
+        );
+        let root = Self::get_root_of_unity(domain_size.try_into().unwrap());
+        let inv_root = root.exp((domain_size as u32 - 1).into());
+        let mut inv_twiddles = Self::get_power_series(inv_root, domain_size / 2);
+        fft::permute(&mut inv_twiddles);
+        inv_twiddles
+    }
+}
+
+impl<const M: u64, const G: u64, const T: u32> FieldElement for BaseElement<M, G, T> {
+    type PositiveInteger = u64;
+    type BaseField = Self;
+
+    const ZERO: Self = Self::new(0);
+    const ONE: Self = Self::new(1);
+
+    const ELEMENT_BYTES: usize = ELEMENT_BYTES;
+
+    const IS_CANONICAL: bool = true;
+
+    fn inv(self) -> Self {
+        Self::new(inv(self.value, M))
+    }
+
+    fn conjugate(&self) -> Self {
+        Self::new(self.value)
+    }
+
+    fn elements_as_bytes(elements: &[Self]) -> &[u8] {
+        let len = elements.len() * Self::ELEMENT_BYTES;
+        let p = elements.as_ptr();
+        unsafe { slice::from_raw_parts(p as *const u8, len) }
+    }
+
+    unsafe fn bytes_as_elements(bytes: &[u8]) -> Result<&[Self], DeserializationError> {
+        if bytes.len() % Self::ELEMENT_BYTES != 0 {
+            return Err(DeserializationError::InvalidValue(format!(
+                "number of bytes ({}) does not divide into whole number of field elements",
+                bytes.len(),
+            )));
+        }
+
+        let p = bytes.as_ptr();
+        let len = bytes.len() / Self::ELEMENT_BYTES;
+
+        if (p as usize) % mem::align_of::<u128>() != 0 {
+            return Err(DeserializationError::InvalidValue(
+                "slice memory alignment is not valid for this field element type".to_string(),
+            ));
+        }
+
+        Ok(slice::from_raw_parts(p as *const Self, len))
+    }
+
+    fn zeroed_vector(n: usize) -> Vec<Self> {
+        // this uses a specialized vector initialization code which requests zero-filled memory
+        // from the OS; here to fulfil trait bounds
+        // We also use u64 to make sure the memory is aligned correctly for our element size.
+        debug_assert_eq!(Self::ELEMENT_BYTES, mem::size_of::<u64>());
+        let result = vec![0u128; n];
+
+        // translate a zero-filled vector of u128s into a vector of base field elements
+        let mut v = core::mem::ManuallyDrop::new(result);
+        let p = v.as_mut_ptr();
+        let len = v.len();
+        let cap = v.capacity();
+        unsafe { Vec::from_raw_parts(p as *mut Self, len, cap) }
+    }
```

Comment:
Since we are not concerned about performance here that much, we can probably just delete this specialized implementation.